### DISCUSSION TITLE
Question: Type flow in a generic validator function

### DISCUSSION THREAD

**Discussion by guitarhero17** [colinhacks/zod#3189] [Q&A] [answered]
Hey guys! I'm pretty new to Zod, I currently have a type issue while trying to create a typed generic validator function. It's about reflecting different status codes coming from an API service. It can receive different `method`s in the request body and delivers appropriate `status` in the response. The idea is to eventually have a generic fetch function, which automatically returns the correctly typed response and an union of all possible responses. As a first step, I'd like to define a generic validator function.

```ts
export const METHOD = {
  ADDRESS_CHECK: 'addressCheck',
  PHONE_CHECK: 'phoneCheck',
  EMAIL_CHECK: 'emailCheck',
  NAME_CHECK: 'nameCheck',
  POSTCODE_AUTOCOMPLETE: 'postCodeAutocomplete',
  CITY_AUTOCOMPLETE: 'cityNameAutocomplete',
  STREET_AUTOCOMPLETE: 'streetAutocomplete',
  FINISH_SESSION: 'doAccounting',
} as const

// getValues() lists all values of METHOD
export const MethodZ = z.enum(getValues(METHOD))

export type Method = z.infer<typeof MethodZ>

// Schemas in the values defined elsewhere (not relevant)
const StatusZ = z.object({
  [METHOD.ADDRESS_CHECK]: AddressCheckStatusSchema,
  [METHOD.PHONE_CHECK]: PhoneCheckStatusSchema,
  [METHOD.EMAIL_CHECK]: EmailCheckStatusSchema,
  [METHOD.NAME_CHECK]: NameCheckStatusSchema,
  [METHOD.POSTCODE_AUTOCOMPLETE]: GeneralStatusSchema,
  [METHOD.CITY_AUTOCOMPLETE]: GeneralStatusSchema,
  [METHOD.STREET_AUTOCOMPLETE]: GeneralStatusSchema,
  [METHOD.FINISH_SESSION]: GeneralStatusSchema,
})

const createStatusObjectSchema = <T extends Method>(method: T) => z.object({
  status: z.array(StatusZ.shape[method]),
})

const NameCheckStatusObjectZ = createStatusObjectSchema(METHOD.NAME_CHECK)

type NameCheckStatusObj = z.infer<typeof NameCheckStatusObjectZ>

// has any[] instead of the appropriate NameCheckStatus[]
type InnerStatus = NameCheckStatusObj['status']
```

Now, If `status` in `createStatusObjectSchema` is just `StatusSchema.shape[method]` (no array) the types flow correctly.

![Screenshot 1](https://github.com/colinhacks/zod/assets/25303218/daed79de-4c0d-48d1-8cb9-3a3c7228b413)

But the API service returns a status array, so I need to have a `z.array(StatusZ.shape[method])`. This messes up the the type and `InnerStatus` is consequently `any[]`. 

![Screenshot 2](https://github.com/colinhacks/zod/assets/25303218/3cae8dfd-a74d-4362-a27c-ae7183c280a0)

I'm pretty sure that I'm missing something here but I can't figure it out on my own.

I'd appreciate any guidance. Thank you!

---------

[Code example](https://stackblitz.com/edit/typescript-3nkb5e?file=schemas%2Fmain.ts)

**Comment by JacobWeisenburger** [comment]
Please send a full reproducible code example.

These variables/functions are missing, so I can't fully understand what you are trying to do:

![image](https://github.com/colinhacks/zod/assets/31667350/87962ba6-3527-4bd5-ae2c-e41cf2a09327)

![image](https://github.com/colinhacks/zod/assets/31667350/5d6950e6-ca6f-4a87-af3e-ad0a9821247b)

**Reply by guitarhero17** [reply]
Hey @JacobWeisenburger , thank you for your reply. I've updated my post with a code example link and some screenshots to better explain what my issue is. Thanks for looking into it.

**Comment by JacobWeisenburger** [comment]
I tried to simplify the example to make it easier to diagnose and everything seems to work fine on this part. So maybe its a problem with your other code.

```ts
const MethodEnum = z.enum( [ 'Foo' ] )
type Method = z.infer<typeof MethodEnum>

const StatusZ = z.object( {
    [ MethodEnum.Enum.Foo ]: z.string()
} )

const createStatusObjectSchema = <T extends Method> ( method: T ) => z.object( {
    status: StatusZ.shape[ method ].array(),
} )

const NameCheckStatusObjectZ = createStatusObjectSchema( MethodEnum.Enum.Foo )

type NameCheckStatusObj = z.infer<typeof NameCheckStatusObjectZ>

type InnerStatus = NameCheckStatusObj[ 'status' ]
// type InnerStatus = string[]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by guitarhero17** [reply]
Hey @JacobWeisenburger ,

I've found that the issue arises as soon as you add other enum members that receive values from diverse types when used as props in the `StatusZ` object.

For example:

```ts
const MethodEnum = z.enum( [ 'Foo', 'Bar' ] )
type Method = z.infer<typeof MethodEnum>

const StatusZ = z.object( {
    [ MethodEnum.Enum.Foo ]: z.string(),
    [ MethodEnum.Enum.Bar ]: z.number()
} )

const createStatusObjectSchema = <T extends Method> ( method: T ) => z.object( {
    status: StatusZ.shape[ method ].array(),
} )

const NameCheckStatusObjectZ = createStatusObjectSchema( MethodEnum.Enum.Bar )

type NameCheckStatusObj = z.infer<typeof NameCheckStatusObjectZ>

type InnerStatus = NameCheckStatusObj[ 'status' ]
// type InnerStatus = string[] | number[]    instead of just   number[]

```

If we define `status` in the z.object returned from `createStatusObjectSchema` as `StatusZ.shape[ method ].array()`, we get `type InnerStatus = string[] | number[]`. So a union of the possible `StatusZ` value types, instead of just  ´number[]`.

For comparison, if we remove `.array()` and just have `StatusZ.shape[ method ]` for the `status` property, the type flows correctly and `type InnerStatus = number`.

Additionally, if `status` is an array but we use the other syntax `z.array(StatusZ.shape[ method ])`, then `type InnerStatus = any[]`, which is confusing for me.

So the problem arises when when trying do have an array. Do you have any ideas why this either gives us a union or `any[]` ?

------
[Updated Code example](https://stackblitz.com/edit/typescript-3nkb5e?file=schemas%2Ftest.ts)

**Comment by JacobWeisenburger** [comment, marked as answer]
In my experience with generic functions, you need to define the return type. Something like this:

```ts
const MethodEnum = z.enum( [ 'Foo', 'Bar' ] )
type Method = z.infer<typeof MethodEnum>

const StatusZ = z.object( {
    [ MethodEnum.Enum.Foo ]: z.string(),
    [ MethodEnum.Enum.Bar ]: z.number()
} )

function createStatusObjectSchema<T extends Method> ( method: T ) {
    return z.object( {
        status: StatusZ.shape[ method ].array(),
    } ) as z.ZodObject<{
        status: z.ZodArray<typeof StatusZ.shape[ typeof method ]>
    }>
}

const fooCheck = createStatusObjectSchema( MethodEnum.Enum.Foo )
type FooStatus = z.infer<typeof fooCheck>[ 'status' ]
// type FooStatus = string[]

const barCheck = createStatusObjectSchema( MethodEnum.Enum.Bar )
type BarStatus = z.infer<typeof barCheck>[ 'status' ]
// type BarStatus = number[]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by guitarhero17** [reply]
Thanks, I ended up completely refactoring my code and changing up the structure. That way I managed to avoid the problem I've described above.
